Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From tsutomu.owa@toshiba.co.jp Mon May 14 10:15:30 2007
Return-Path: <tsutomu.owa@toshiba.co.jp>
X-Spam-Checker-Version: SpamAssassin 3.1.7-deb (2006-10-05) on debian
X-Spam-Level: 
X-Spam-Status: No, score=0.0 required=5.0 tests=UNPARSEABLE_RELAY 
	autolearn=ham version=3.1.7-deb
Received: from imx12.toshiba.co.jp (imx12.toshiba.co.jp [61.202.160.132])
	by mail.tglx.de (Postfix) with ESMTP id 7006365C065 for
	<tglx@linutronix.de>; Mon, 14 May 2007 10:15:30 +0200 (CEST)
Received: from wall11.toshiba.co.jp (wall11 [133.199.90.149]) by
	imx12.toshiba.co.jp  with ESMTP id l4E8FKmi007480; Mon, 14 May 2007
	17:15:20 +0900 (JST)
Received: (from root@localhost) by wall11.toshiba.co.jp  id l4E8FKaH003434;
	Mon, 14 May 2007 17:15:20 +0900 (JST)
Received: from ovp11.toshiba.co.jp [133.199.90.148]  by
	wall11.toshiba.co.jp with ESMTP id TAA03430; Mon, 14 May 2007 17:15:20 +0900
Received: from mx2.toshiba.co.jp (localhost [127.0.0.1]) by
	ovp11.toshiba.co.jp  with ESMTP id l4E8FJCq025717; Mon, 14 May 2007
	17:15:19 +0900 (JST)
Received: from rdcgw.rdc.toshiba.co.jp by toshiba.co.jp id l4E8FJ3Y013473;
	Mon, 14 May 2007 17:15:19 +0900 (JST)
Received: from island.swc.toshiba.co.jp by rdcgw.rdc.toshiba.co.jp
	(8.8.8p2+Sun/3.7W) with ESMTP id RAA01521; Mon, 14 May 2007 17:15:18 +0900
	(JST)
Received: from forest.toshiba.co.jp (forest [133.196.122.2]) by
	island.swc.toshiba.co.jp (Postfix) with ESMTP id 87FCB40002; Mon, 14 May
	2007 17:15:10 +0900 (JST)
Date: Mon, 14 May 2007 17:15:10 +0900
Message-ID: <yyir6pj95yp.wl@toshiba.co.jp>
From: Tsutomu OWA <tsutomu.owa@toshiba.co.jp>
To: linuxppc-dev@ozlabs.org, linux-kernel@vger.kernel.org
Cc: mingo@elte.hu, tglx@linutronix.de
Subject: Re: [patch 1/5] powerpc 2.6.21-rt1: add mcount() and _mcount()
In-Reply-To: <yyisl9z97hg.wl@toshiba.co.jp>
References: <yyisl9z97hg.wl@toshiba.co.jp>
User-Agent: Wanderlust/2.8.1 (Something) Emacs/20.7 Mule/4.0 (HANANOEN)
Organization: Software Engineering Center, TOSHIBA.
MIME-Version: 1.0 (generated by SEMI 1.14.4 - "Hosorogi")
Content-Type: text/plain; charset=US-ASCII
X-Evolution-Source: imap://tglx%40linutronix.de@localhost:8993/
Content-Transfer-Encoding: 8bit


add mcount() and _mcount() for latency trace support.

Signed-off-by: Tsutomu OWA <tsutomu.owa@toshiba.co.jp>
-- owa

---
 arch/powerpc/kernel/entry_64.S |   60 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 60 insertions(+)

Index: linux-2.6.24-rc2-rt1/arch/powerpc/kernel/entry_64.S
===================================================================
--- linux-2.6.24-rc2-rt1.orig/arch/powerpc/kernel/entry_64.S
+++ linux-2.6.24-rc2-rt1/arch/powerpc/kernel/entry_64.S
@@ -846,3 +846,63 @@ _GLOBAL(enter_prom)
 	ld	r0,16(r1)
 	mtlr    r0
         blr
+
+#ifdef CONFIG_MCOUNT
+/*
+ * code almost taken from entry_32.S
+ */
+#define MCOUNT_FRAME_SIZE 32
+_GLOBAL(mcount)
+	stdu	r1,-MCOUNT_FRAME_SIZE(r1)
+	mflr	r3
+
+	LOAD_REG_ADDR(r5,mcount_enabled)
+	lwz	r5,0(r5)
+	std	r3,MCOUNT_FRAME_SIZE+16(r1)
+	cmpwi	r5,0
+	beq	1f
+
+	/* r3 contains lr (eip), put parent lr (parent_eip) in r4 */
+	ld	r4,MCOUNT_FRAME_SIZE(r1)
+	ld	r4,16(r4)
+	bl	.__trace
+	nop
+1:
+	ld	r0,MCOUNT_FRAME_SIZE+16(r1)
+	mtlr	r0
+	addi	r1,r1,MCOUNT_FRAME_SIZE
+	blr
+
+/*
+ * Based on glibc-2.4/sysdeps/powerpc/powerpc64/ppc-mcount.S
+ *
+ * We don't need to save the parameter-passing registers as gcc takes
+ * care of that for us.  Thus this function looks fairly normal.
+ * In fact, the generic code would work for us.
+ */
+_GLOBAL(_mcount)
+	/* return if we're in real mode. */
+	mfmsr	r3
+	andi.	r0,r3,MSR_IR|MSR_DR		/* see if relocation is on? */
+	beqlr					/* if not, do nothing. */
+	/* we're in translation mode. keep going. */
+	mflr	r3
+	ld	r11,0(r1)			/* load back chain ptr */
+	stdu	r1,-STACK_FRAME_OVERHEAD(r1)
+	std	r3,STACK_FRAME_OVERHEAD+16(r1)
+	ld	r4,16(r11)			/* LR in back chain */
+	LOAD_REG_ADDR(r5,mcount_enabled)
+	lwz	r5,0(r5)
+	cmpwi	r5,0				/* see if mcount_enabled? */
+	beq	1f				/* if disabled, then skip */
+
+	/* r3 contains lr (eip), put parent lr (parent_eip) in r4 */
+	bl	 .__trace
+	nop
+1:
+	ld	r0,STACK_FRAME_OVERHEAD+16(r1)	/* restore saved LR */
+	mtlr	r0
+	addi	r1,r1,STACK_FRAME_OVERHEAD
+	blr
+
+#endif /* CONFIG_MCOUNT */
